### PR TITLE
Move permanence from effect instance to effect type

### PR DESCRIPTION
**Pull request by Coolthulhu**
The code was bloated with needlessly explicit "is this effect permanent?", when in vast majority of cases, the answer was "if its type is".

This will break some mods, but the fix is easy: just declare the effect type as `"permanent": true`.

No new functionality, just less garbage in code.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -5973,6 +5951,12 @@ bool Character::is_immune_effect( const efftype_id &eff ) const
         return is_immune_damage( DT_ACID ) || has_trait( trait_SLIMY ) || has_trait( trait_VISCOUS );
     } else if( eff == effect_nausea ) {
         return has_trait( trait_STRONGSTOMACH );
+    } else if( eff == effect_bleed ) {
+        return std::any_of( cached_mutations.begin(),
+        cached_mutations.end(), []( const mutation_branch * mut ) {
+            // Ugly, it was badly implemente and should be a flag
+            return mut->bleed_resist > 0;
+        } );
```
Comment (olanti-p) [inline src/character.cpp:5959]: `cached_mutations` does not include mutations from enchantments

File: src/effect.h [hunk 2]
```diff
@@ -76,6 +76,9 @@ class effect_type
 
         bool is_show_in_info() const;
 
+        /** Does the effect  */
+        bool is_permanent() const;
```
Comment (olanti-p) [inline src/effect.h:80]: Missed a word?

File: src/creature.h [hunk 3]
```diff
@@ -335,12 +335,18 @@ class Creature
...
+                                 int intensity = 0, bool force = false, bool deferred = false );
         /** Gives chance to save via environmental resist, returns false if resistance was successful. */
+        bool add_env_effect( const efftype_id &eff_id, body_part vector, bool REMOVED, int strength,
+                             const time_duration &dur,
+                             body_part bp = num_bp, int intensity = 1,
+                             bool force = false ) = delete;
         bool add_env_effect( const efftype_id &eff_id, body_part vector, int strength,
                              const time_duration &dur,
-                             body_part bp = num_bp, bool permanent = false, int intensity = 1,
+                             body_part bp = num_bp, int intensity = 1,
```
Comment (olanti-p) [inline src/creature.h:349]: Was REMOVED argument here intended to be 3rd, not 6th?

File: src/character.cpp (hunk 1 above)
Comment (Coolthulhu) [reply src/character.cpp:5959]: Maybe it should. After all, it's supposed to contain all the mutations that exist for "mechanical purposes".
For example, night vision, social mods, hp modifier, dodge modifier - all of those check `cached_mutations`.

File: src/creature.h (hunk 3 above)
Comment (Coolthulhu) [reply src/creature.h:349]: Oh right, this one has different syntax. `add_env_effect` is such an artifact of old times.

File: src/character.cpp (hunk 1 above)
Comment (Coolthulhu) [reply src/character.cpp]: Alright, I made it use `mutation_value`. Now even though it still doesn't respect uncached mutations, it is only "just as wrong as" the other effects and will automatically be fixed when a proper solution is introduced.

File: src/npctalk.cpp [hunk 4]
```diff
@@ -1918,24 +1918,29 @@ void talk_effect_fun_t::set_add_effect( const JsonObject &jo, const std::string
...
         const std::string dur_string = jo.get_string( "duration" );
         if( dur_string == "PERMANENT" ) {
-            permanent = true;
+            // This is immensely ugly, we need json.just_warn_with_context
+            try {
+                jo.throw_error( "Effect permanence has been moved to effect_type. Set permanence there." );
+                duration = 365_days;
+            } catch( const JsonError &e ) {
+                debugmsg( "\n%s", e.what() );
+            }
```
Comment (olanti-p) [inline src/npctalk.cpp:1931]: `duration = 365_days;` is unreachable here.

Also, `throw_error` supports `name` argument, supplying "duration" should result is more precise location:
```
 "effect": { "u_add_effect": "u_met_Jenny_Forcette", "duration":
                                                                ^
                                                                 "PERMANENT" },
```

File: src/magic_spell_effect.cpp [hunk 5]
```diff
@@ -403,7 +403,7 @@ static void add_effect_to_target( const tripoint &target, const spell &sp )
     if( guy ) {
         for( const body_part bp : all_body_parts ) {
             if( sp.bp_is_affected( bp ) ) {
-                guy->add_effect( spell_effect, dur_td, bp, sp.has_flag( spell_flag::PERMANENT ) );
+                guy->add_effect( spell_effect, dur_td, bp );
```
Comment (olanti-p) [inline src/magic_spell_effect.cpp:406]: If spell duration is not specified, effect duration here defaults to 0.
Permanent effects with 0 duration get removed after 1 turn.
This is somewhat counter-intuitive... Maybe check on finalize, or limit here (and 6 lines below) to min. 1 second?

**Pull request merged**